staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,232 @@
+const (
+	// FlowDistinguisherMethodByUser works by computing flow distinguisher with the user name in the request context so
+	// that the requests from the same user will enqueued into the same queue for processing.
+	FlowDistinguisherMethodByUser  FlowDistinguisherMethodType = "ByUser"
+	// FlowDistinguisherMethodByNamespace works by computing flow distinguisher with the user requested namespace. Any
+	// user requesting the namespace will be put into the same queue for processing.
+	FlowDistinguisherMethodByNamespace FlowDistinguisherMethodType = "ByNamespace"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchema defines the schema of a group of flows. Note that a flow makes up of a set of inbound API requests with
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// PriorityLevelConfigurationRef can reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfiguration PriorityLevelConfigurationReference `json:"priorityLevelConfiguration,omitempty" protobuf:"bytes,1,opt,name=priorityLevelConfiguration"`
+	// MatchingPrecedence is used to choose among the FlowSchemas that match a given request.  The chosen
+	// FlowSchema is among those with the numerically lowest (which we take to be logically highest)
+	// MatchingPrecedence.  Each MatchingPrecedence value must be non-negative.
```

Comment:
I'd suggest a range or give examples of values that make sense.

Reply:
Authors need to know the MatchingPrecedence values of the default objects.

We probably do not want to list the default objects in this comment.  What is the right way to direct readers to a listig of the default objects?

Reply:
ideally, i think we should unify/converge the usages of priorities in k/k apis e.g. apiservices. and the apis will share a common link to the docs describing how the priority work.

Reply:
My point here is that an administrator creating or modifying FlowSchemaSpec objects needs to know not only general guidelines about matching precendence values but also the specific precendence values of the other existing FlowSchema objects.  Actually, that is what matters, not the precedence values of the default objects.  So really, an administrator has to have done a list of the other objects before she can proceed to make any changes.  Given that, I do not think we even need general guidelines --- the specifics of the other existing objects should suffice.